pkg/solana/logpoller/job.go

```
@@ -0,0 +1,145 @@
+type processEventJob struct {
+	parser ProgramEventProcessor
+	event  ProgramEvent
+}
+
+func (j *processEventJob) String() string {
+	return fmt.Sprintf("processEventJob for event: %s", j.event.FunctionName)
+}
+
+func (j *processEventJob) Run(_ context.Context) error {
+	return j.parser.Process(j.event)
+}
+
+// getTransactionsFromBlockJob is a job that fetches transaction signatures from a block and loads
+// the job queue with getTransactionLogsJobs for each transaction found in the block.
+type getTransactionsFromBlockJob struct {
+	slotNumber uint64
+	client     RPCClient
+	parser     ProgramEventProcessor
+	chJobs     chan Job
+}
+
+func (j *getTransactionsFromBlockJob) String() string {
+	return fmt.Sprintf("getTransactionsFromBlockJob for block: %d", j.slotNumber)
+}
+
+func (j *getTransactionsFromBlockJob) Run(ctx context.Context) error {
+	var excludeRewards bool
+
+	block, err := j.client.GetBlockWithOpts(
+		ctx,
+		j.slotNumber,
+		// You can specify more options here:
+		&rpc.GetBlockOpts{
+			Encoding:   solana.EncodingBase64,
+			Commitment: rpc.CommitmentFinalized,
+			// get the full transaction details
+			TransactionDetails: rpc.TransactionDetailsFull,
+			// exclude rewards
+			Rewards: &excludeRewards,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	blockSigsOnly, err := j.client.GetBlockWithOpts(
+		ctx,
+		j.slotNumber,
+		// You can specify more options here:
+		&rpc.GetBlockOpts{
+			Encoding:   solana.EncodingBase64,
+			Commitment: rpc.CommitmentFinalized,
+			// get the signatures only
+			TransactionDetails: rpc.TransactionDetailsSignatures,
+			// exclude rewards
+			Rewards: &excludeRewards,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	detail := eventDetail{
+		blockHash: block.Blockhash,
+	}
+
+	if block.BlockHeight != nil {
+		detail.blockNumber = *block.BlockHeight
+	}
+
+	for idx, trx := range block.Transactions {
+		detail.trxIdx = idx
+		if len(blockSigsOnly.Signatures)-1 <= idx {
```

Is it possible that a transaction does not have a corresponding signature?

---

The integration test using a test validator node implies that each transaction should have a signature. At this point I can only assume that live nodes should function the same way.

---

nit: should we throw an error `if len(block.Transactions) != len(blockSigsOnly.Signatures)`

---

Seems fair